### DISCUSSION TITLE
Updating multiple stamps for one resolver

### DISCUSSION THREAD

**Discussion by ghost** [DNSCrypt/dnscrypt-proxy#2422] [Potential issues]
A few resolvers have multiple stamps for one host.
I just saw `[INFO] Updating stamp for` for such a host where just one stamp was replaced by another one (without the stamps being recently changed). Is this intended behaviour? It just looks to me as updating stamps doesn't consider the possibility of multiple stamps for one host.

**Comment by lifenjoiner** [comment]
I guess it is intentional behavior.

Think about DoH. A domain could be resolved into several IPs. Clients (browsers) picks one to communicate with, normally. Here is the same.
And then, consider [Load Balancing](https://github.com/DNSCrypt/dnscrypt-proxy/wiki/Load-Balancing-Options):
1. A client chooses several fastest severs, by `lb_strategy`. It needs the servers are: sorted ==> measured ==> consistent IP.
2. The sever hopes the clients to be dispatched, to IPs.
3. IPs are not with the same availability to clients. Periodic refreshing should give a chance to users to get a valid IP, if the previous one is low quality.

So now, I will say it is a load balancing strategy.

Further more, there seems to be a flaw in the rotation:
It always replaces the old one with the 1st new found. That is rotating between the first two.
PR https://github.com/DNSCrypt/dnscrypt-proxy/pull/2426 for this.

**Comment by ghost** [comment]
Sorry for explaining poorly.
I'm talking about when `public-resolvers.md` is updated. There it seems to do things like changing {stamp1, stamp2} to {stamp**2**, stamp2} if I understand correctly.

**Reply by lifenjoiner** [reply]
The adopted servers are stored by their names.
That means, if group (some providers provide groups of service) A has 2 stamps, say stamp1 and stamp2:
when refreshing, stamp1 will be replaced with stamp2, or stamp2 will be replaced with stamp1, they won't be adopted at the same time.

Do you have logs? That will make it clear.

**Reply by ghost** [reply]
Ah, I thought that all stamps would be adopted at once and DNSCrypt chooses when it makes a query, my bad.

The only question that remains is what's better, using one provider name with several stamps or giving every stamp a unique provider name?
